File: backend/app/services/user_service.py
import uuid
from typing import List, Optional, Tuple
from sqlalchemy.ext.asyncio import AsyncSession

from app.audit.service import AuditService
from app.core.exceptions import BadRequestException, ConflictException, ForbiddenException, NotFoundException
from app.models.user import User, UserRole
from app.repositories.user_repository import UserRepository
from app.schemas.user import UserCreate, UserUpdate, UserChangePassword
from app.security.password import hash_password, verify_password


class UserService:
    def __init__(self, db: AsyncSession):
        self.db = db
        self.repo = UserRepository(db)
        self.audit = AuditService(db)

    async def get_by_id(self, user_id: uuid.UUID) -> User:
        user = await self.repo.get_by_id(user_id)
        if not user:
            raise NotFoundException("User", user_id)
        return user

    async def list(
        self,
        page: int = 1,
        page_size: int = 20,
    ) -> Tuple[List[User], int]:
        return await self.repo.get_all(
            offset=(page - 1) * page_size,
            limit=page_size,
        )

    async def create(self, data: UserCreate, current_user: User) -> User:
        if current_user.role != UserRole.ADMIN:
            raise ForbiddenException("Only admins can create users")
        if await self.repo.get_by_email(data.email):
            raise ConflictException(f"Email '{data.email}' is already registered")

        user = await self.repo.create(
            {
                "full_name": data.full_name,
                "email": data.email,
                "hashed_password": hash_password(data.password),
                "role": data.role,
                "is_active": True,
                "is_verified": True,
            }
        )
        await self.audit.log(
            action="USER_CREATE",
            resource="users",
            resource_id=str(user.id),
            user_id=current_user.id,
            new_values={"email": user.email, "role": user.role.value},
        )
        return user

    async def update(self, user_id: uuid.UUID, data: UserUpdate, current_user: User) -> User:
        user = await self.get_by_id(user_id)
        updated = await self.repo.update(user, data.model_dump(exclude_none=True))
        await self.audit.log(
            action="USER_UPDATE",
            resource="users",
            resource_id=str(user_id),
            user_id=current_user.id,
            new_values=data.model_dump(exclude_none=True),
        )
        return updated

    async def delete(self, user_id: uuid.UUID, current_user: User) -> None:
        if user_id == current_user.id:
            raise BadRequestException("Cannot delete your own account")
        user = await self.get_by_id(user_id)
        await self.repo.soft_delete(user)
        await self.audit.log(
            action="USER_DELETE",
            resource="users",
            resource_id=str(user_id),
            user_id=current_user.id,
        )

    async def change_password(
        self, user_id: uuid.UUID, data: UserChangePassword, current_user: User
    ) -> User:
        user = await self.get_by_id(user_id)
        if not verify_password(data.current_password, user.hashed_password):
            raise BadRequestException("Current password is incorrect")
        updated = await self.repo.update(
            user, {"hashed_password": hash_password(data.new_password)}
        )
        return updated

File: frontend/src/services/user.service.ts
import { apiClient } from '@/lib/axios'

export interface User { id: string; full_name: string; email: string; role: string; is_active: boolean; is_verified: boolean; created_at: string }
export interface UsersResponse { data: User[]; total: number; page: number; page_size: number; total_pages: number }
export interface CreateUserPayload { full_name: string; email: string; password: string; role: string }

export const userService = {
  list: (params?: { page?: number; page_size?: number }) =>
    apiClient.get<UsersResponse>('/users', { params }).then((r) => r.data),

  get: (id: string) =>
    apiClient.get<{ data: User }>(`/users/${id}`).then((r) => r.data.data),

  create: (data: CreateUserPayload) =>
    apiClient.post<{ data: User }>('/users', data).then((r) => r.data.data),

  update: (id: string, data: Partial<{ full_name: string; role: string; is_active: boolean }>) =>
    apiClient.put<{ data: User }>(`/users/${id}`, data).then((r) => r.data.data),

  delete: (id: string) => apiClient.delete(`/users/${id}`),

  changePassword: (current_password: string, new_password: string) =>
    apiClient.post('/users/me/change-password', { current_password, new_password }),
}
